src/libProtoServer/Server.cpp

```diff
@@ -0,0 +1,1215 @@
+    }
+    // for the latest block
+    TxBlockHeader txhead =
+        m_mediator.m_txBlockChain.GetBlock(currBlockNum).GetHeader();
+    SHA2<HASH_TYPE::HASH_VARIANT_256> sha2;
+    vector<unsigned char> vec;
+    txhead.Serialize(vec, 0);
+    sha2.Update(vec);
+    const vector<unsigned char>& resVec = sha2.Finalize();
+
+    m_TxBlockCache.second.insert_new(m_TxBlockCache.second.size(),
+                                     DataConversion::Uint8VecToHexStr(resVec));
+    m_TxBlockCache.first = currBlockNum;
+  }
+
+  unsigned int offset = PAGE_SIZE * (page - 1);
+  if (page <= NUM_PAGES_CACHE) {  // can use cache
+    boost::multiprecision::uint256_t cacheSize(m_TxBlockCache.second.capacity());
+
+    if (cacheSize > m_TxBlockCache.second.size()) {
+      cacheSize = m_TxBlockCache.second.size();
+    }
+
+    uint64_t size = m_TxBlockCache.second.size();
+    for (unsigned int i = offset; i < PAGE_SIZE + offset && i < cacheSize; i++) {
+      auto blockData = ret.add_data();
+      blockData->set_hash(m_TxBlockCache.second[size - i - 1]);
+      blockData->set_blocknum(int(currBlockNum - i));
+    }
+
+  } else {
+    for (uint64_t i = offset; i < PAGE_SIZE + offset && i <= currBlockNum; i++) {
+      auto blockData = ret.add_data();
+      blockData->set_hash(m_mediator.m_txBlockChain.GetBlock(currBlockNum - i + 1).GetHeader().GetPrevHash().hex());
+      blockData->set_blocknum(int(currBlockNum - i));
+    }
+  }
+
+  return ret;
+}
+
+
+ProtoBlockChainInfo Server::GetBlockchainInfo() {
+  ProtoBlockChainInfo ret;
+
+  ret.set_numpeers(Server::GetNumPeers().result());
+  ret.set_numtxblocks(Server::GetNumTxBlocks().result());
+  ret.set_numdsblocks(Server::GetNumDSBlocks().result());
+  ret.set_numtxns(Server::GetNumTransactions().result());
+  ret.set_txrate(Server::GetTransactionRate().result());
+  ret.set_txblockrate(Server::GetTxBlockRate().result());
+  ret.set_dsblockrate(Server::GetDSBlockRate().result());
+  ret.set_currentminiepoch(Server::GetCurrentMiniEpoch().result());
+  ret.set_currentdsepoch(Server::GetCurrentDSEpoch().result());
+  ret.set_numtxnsdsepoch(Server::GetNumTxnsDSEpoch().result());
+  ret.set_numtxnstxepoch(Server::GetNumTxnsTxEpoch().result());
+
+  ProtoShardingStruct sharding = Server::GetShardingStructure();
+  ret.set_allocated_shardingstructure(&sharding);
+
+  return ret;
+}
+
+
+ProtoTxHashes Server::GetRecentTransactions() {
+  LOG_MARKER();
+
+  lock_guard<mutex> g(m_mutexRecentTxns);
+
+  uint64_t actualSize(m_RecentTransactions.capacity());
+  if (actualSize > m_RecentTransactions.size()) {
+    actualSize = m_RecentTransactions.size();
+  }
+
+  ProtoTxHashes ret;
+  ret.set_number(int(actualSize));
+
+  uint64_t size = m_RecentTransactions.size();
+  for (uint64_t i = 0; i < actualSize; i++) {
+    auto txhash = ret.add_txhashes();
+    txhash->set_txhash(m_RecentTransactions[size - i - 1]);
+  }
+
+  return ret;
+}
+
+
+void Server::AddToRecentTransactions(ProtoTxHash& protoTxHash) {
```

Comment:
dont need to change this one, this one is auxillary function

Reply:
Okay. Removing protobuf from it.